build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1188,6 +1212,34 @@ message UpdateActionResultRequest {
   ResultsCachePolicy results_cache_policy = 4;
 }
 
+// A request message for
+// [RepositoryCache.Download][build.bazel.remote.execution.v2.RepositoryCache.Download].
+message DownloadRequest {
+  // The instance of the execution system to operate against. A server may
+  // support multiple instances of the execution system (with their own workers,
+  // storage, caches, etc.). The server MAY require use of this field to select
+  // between them in an implementation-defined fashion, otherwise it can be
+  // omitted.
+  string instance_name = 1;
+
+  // List of mirror URLs referencing the same file.
+  repeated string urls = 2;
```

Comment:
Open question: Should this be URLs or URIs? Or either? Aka should this be 'https://some.host/some.resource', or 'some.host/some.resource'? 

On one hand, not all URIs will be URLs - git repos are an example where the protocol may reasonable be decoupled from the identifier. And if one knows the host, URIs should theoretically be easily mapped to URLs. On the other hand, if you want to support truly host-agnostic fetching, URLs may be required, or we'll just end up with ad-hoc standards like "for unknown hosts, HTTPS is assumed".

I'm leaning towards saying "URL" is correct, and servers may rewrite it if needed. But I'd be interested in what you all think.

Reply:
Actually, perhaps neither URL nor URI are broad enough? A file can be expressed by a URL. A git repo is probably repo+ref. Either might be augmented by a minimum fetch timestamp. A monorepo CL is potentially just a CL, but in reality could be some complicated combination of CL, branch, diffs, workspace, and who knows what else. And then a file *in* a repo is a path on top of one of those.

I'm torn between saying that a URI can encode any packed data it wants, and so suffices for all of those use-cases, and that we'll just end up inventing ugly and broken packing schemes if we go that route, making strongly typed protos preferred.

I'd be interested in a client perspective - for Bazel or others, which would be more usable in practice? Typed protos would probably require you to know the types you want to support, but maybe you need to anyways; untyped URIs probably require you to get those directly from clients, or embed URI-encoding logic. I don't know which is worse.

Reply:
I think URLs. A client that requests `https://example.com/some/archive.tar.gz` is being clear about what it expects, and all of the obvious ways to download a file (HTTP, FTP, S3/GCS, Git, SCP) either identify resources by URL natively or are trivially adaptable to do so.

The `RepositoryCache.Download()` RPC concerns itself only with downloading single files. Downloading of multi-file content such a directory trees is out of scope -- so we don't need to solve the issue of encoding arbitrary revision identifiers into a URL. If there is some future person who would like to identify downloadable content by diff, then that person will have enough context on their use case to write a reasonable API.

Reply:
From a client perspective, I think URL-encoding would be delegated to Starlark, which avoids requiring Bazel to be aware of the possible formats. +1 for URLs.